x/emissions/keeper/keeper.go

```diff
@@ -3537,3 +3682,263 @@ func (k *Keeper) GetPreviousForecasterScoreRatio(ctx context.Context, topicId To
 	}
 	return forecastTau, nil
 }
+
+// AddActiveInferer adds an inferer to the active inferers set for a topic
+func (k *Keeper) AddActiveInferer(ctx context.Context, topicId TopicId, inferer ActorId) error {
+	if err := types.ValidateTopicId(topicId); err != nil {
+		return errorsmod.Wrap(err, "invalid topic id")
+	}
+	if err := types.ValidateBech32(inferer); err != nil {
+		return errorsmod.Wrap(err, "invalid inferer address")
+	}
+	key := collections.Join(topicId, inferer)
+	return k.activeInferers.Set(ctx, key)
+}
+
+// IsActiveInferer checks if an inferer is in the active inferers set for a topic
+func (k *Keeper) IsActiveInferer(ctx context.Context, topicId TopicId, inferer ActorId) (bool, error) {
+	key := collections.Join(topicId, inferer)
+	return k.activeInferers.Has(ctx, key)
+}
+
+// RemoveActiveInferer removes an inferer from the active inferers set for a topic
+func (k *Keeper) RemoveActiveInferer(ctx context.Context, topicId TopicId, inferer ActorId) error {
+	key := collections.Join(topicId, inferer)
+	return k.activeInferers.Remove(ctx, key)
+}
+
+// GetActiveInferersForTopic returns all active inferers for a specific topic
+func (k *Keeper) GetActiveInferersForTopic(ctx context.Context, topicId TopicId) ([]ActorId, error) {
+	var inferers []ActorId
+	rng := collections.NewPrefixedPairRange[TopicId, ActorId](topicId)
+	err := k.activeInferers.Walk(ctx, rng, func(key collections.Pair[TopicId, ActorId]) (bool, error) {
+		inferers = append(inferers, key.K2())
+		return false, nil
+	})
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "error walking active inferers")
+	}
+	return inferers, nil
+}
+
+// AddActiveForecaster adds a forecaster to the active forecasters set for a topic
+func (k *Keeper) AddActiveForecaster(ctx context.Context, topicId TopicId, forecaster ActorId) error {
+	if err := types.ValidateTopicId(topicId); err != nil {
+		return errorsmod.Wrap(err, "invalid topic id")
+	}
+	if err := types.ValidateBech32(forecaster); err != nil {
+		return errorsmod.Wrap(err, "invalid forecaster address")
+	}
+	key := collections.Join(topicId, forecaster)
+	return k.activeForecasters.Set(ctx, key)
+}
+
+// IsActiveForecaster checks if a forecaster is in the active forecasters set for a topic
+func (k *Keeper) IsActiveForecaster(ctx context.Context, topicId TopicId, forecaster ActorId) (bool, error) {
+	key := collections.Join(topicId, forecaster)
+	return k.activeForecasters.Has(ctx, key)
+}
+
+// RemoveActiveForecaster removes a forecaster from the active forecasters set for a topic
+func (k *Keeper) RemoveActiveForecaster(ctx context.Context, topicId TopicId, forecaster ActorId) error {
+	key := collections.Join(topicId, forecaster)
+	return k.activeForecasters.Remove(ctx, key)
+}
```

Comment:
Should this not be indexed by nonce as well? because active set changes across nonces... Not every worker is active for every epoch. An active participant may not even participate in all epochs. Also, notably, an epoch can be judged way later (ground truth lag blocks later) so if there is a dynamic active set (which there is) would the wrong actors be scored when an earlier nonce is ready to be scored?

Reply:
oh I"m sorry this is wrong -- you don't query this map when scoring workers. These maps are reset between epochs per actor type (Worker | Reputer)